File: hosts/capacitor/ios/App/NativeEntitlementSession/NativeEntitlementSession.swift
import Foundation

struct DurableOwnership: Equatable {
    let doubleInfinityPoints: Bool
    let developerOptions: Bool
    let supporterCatGallery: Bool
}

/**
 * Process-local authority for the latest successful Store ownership snapshot.
 * Keychain state remains the offline fallback, but cannot overrule a newer
 * live result from StoreKit during the same process lifetime.
 */
final class NativeEntitlementSession {
    private var authoritativeOwnership: DurableOwnership?
    private var pendingPersistence: DurableOwnership?

    func resolve(
        persistedProviderOwnership: DurableOwnership,
        legacyDoubleInfinityPoints: Bool
    ) -> DurableOwnership {
        let provider = authoritativeOwnership ?? persistedProviderOwnership
        return DurableOwnership(
            doubleInfinityPoints:
                provider.doubleInfinityPoints || legacyDoubleInfinityPoints,
            developerOptions: provider.developerOptions,
            supporterCatGallery:
                provider.supporterCatGallery ||
                    persistedProviderOwnership.supporterCatGallery
        )
    }

    func applyProviderOwnership(
        _ ownership: DurableOwnership,
        persist: (DurableOwnership) -> Bool
    ) -> Bool {
        authoritativeOwnership = ownership
        pendingPersistence = ownership
        return persistPending(using: persist)
    }

    func retryPendingPersistence(
        persist: (DurableOwnership) -> Bool
    ) -> Bool {
        guard pendingPersistence != nil else { return true }
        return persistPending(using: persist)
    }

    private func persistPending(
        using persist: (DurableOwnership) -> Bool
    ) -> Bool {
        guard let pending = pendingPersistence else { return true }
        guard persist(pending) else { return false }
        if pendingPersistence == pending { pendingPersistence = nil }
        return true
    }
}

File: hosts/capacitor/android/app/src/main/java/com/blindsidedgames/idledysonswarm/NativeEntitlementSession.kt
package com.blindsidedgames.idledysonswarm

internal data class DurableOwnership(
    val doubleInfinityPoints: Boolean,
    val developerOptions: Boolean,
    val supporterCatGallery: Boolean,
)

/**
 * Process-local authority for the latest successful Store ownership snapshot.
 * Disk remains the offline fallback, but cannot overrule a newer live result.
 */
internal class NativeEntitlementSession {
    private var authoritativeOwnership: DurableOwnership? = null
    private var pendingPersistence: DurableOwnership? = null

    fun resolve(
        persistedProviderOwnership: DurableOwnership,
        legacyDoubleInfinityPoints: Boolean,
    ): DurableOwnership {
        val provider = authoritativeOwnership ?: persistedProviderOwnership
        return DurableOwnership(
            doubleInfinityPoints =
                provider.doubleInfinityPoints || legacyDoubleInfinityPoints,
            developerOptions = provider.developerOptions,
            supporterCatGallery =
                provider.supporterCatGallery ||
                    persistedProviderOwnership.supporterCatGallery,
        )
    }

    fun applyProviderOwnership(
        ownership: DurableOwnership,
        persist: (DurableOwnership) -> Boolean,
    ): Boolean {
        authoritativeOwnership = ownership
        pendingPersistence = ownership
        return persistPending(persist)
    }

    fun retryPendingPersistence(
        persist: (DurableOwnership) -> Boolean,
    ): Boolean = pendingPersistence?.let { persistPending(persist) } ?: true

    private fun persistPending(
        persist: (DurableOwnership) -> Boolean,
    ): Boolean {
        val pending = pendingPersistence ?: return true
        if (!persist(pending)) return false
        if (pendingPersistence == pending) pendingPersistence = null
        return true
    }
}
